core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2251,6 +2251,27 @@ public void setNodes(final List<? extends Node> n) throws IOException {
         }
     }
 
+    /**
+     * Reset labels and remove invalid ones for the given nodes.
+     * @param nodes the nodes taken as reference to update labels
+     */
+    void trimLabels(Node... nodes) {
+        Set<LabelAtom> includedLabels = new HashSet<>();
+        for (Node n : nodes) {
+            includedLabels.addAll(n.getAssignedLabels());
+        }
+        Set<Label> nodeLabels = new HashSet<>(this.getAssignedLabels());
+        this.getNodes().forEach(n -> nodeLabels.addAll(n.getAssignedLabels()));
+        for (Iterator<Label> itr = this.labels.values().stream().filter(l -> includedLabels.contains(l)).iterator(); itr.hasNext();) {
+            Label l = itr.next();
+            if (nodeLabels.contains(l) || this.clouds.stream().anyMatch(c -> c.canProvision(l))) {
```

Comment:
rather than calling `c.canProvision()`  can we check if the label has any clouds listed?  (and avoind asking if a cloud can provision again)?
`! l.getClouds().isEmpty()`  (if the set is not already computed it will compute it - but that is pretty much what we do here anyway - but at least this gives us a possiblity of avoiding a call?

Reply:
Applied